document cluster taint management (#843)
karmada-io/website: docs/userguide/failover/cluster-taint-management.md

```

Even if users configure `--allow-no-execute-taint-policy=true` in the `karmada-webhook`, it does not guarantee that the system will act on **NoExecute** taints. Users must still pay attention to the following configurations in the `karmada-controller-manager` component:

- karmada-controller-manager：
```
--enable-no-execute-taint-eviction bool: Enables controller response to NoExecute taints on clusters, which triggers eviction of workloads without explicit tolerations. Given the impact of eviction caused by NoExecute Taint.
 This parameter is designed to remain disabled by default and requires careful evaluation by administrators before being enabled.
```

### Failover FeatureGate

The cluster taint management capability requires the **Failover FeatureGate** to be enabled. **Failover FeatureGate** is currently in the **Beta** stage and is turned off by default. If you wish to enable automatic taint management for clusters via ClusterTaintPolicy, you can configure it in the `karmada-controller-manager` as follows:

```
--feature-gates=Failover=true
```

> ## Cluster Taints Tolerations

Author: Use this one: Cluster Taint Tolerations

---
title: Cluster Taint Management
---

> **Cluster Taint Mechanism for Application Scheduling and Migration**
> 
> The Karmada system employs cluster taints to:
> 1. Influence scheduling decisions for new applications targeting specific clusters.
> 2. Trigger the migration of existing applications running on those clusters when conditions require.
> 
> Taint Behavior Examples: When a cluster is tainted with the `NoExecute` effect:
> - Applications propagated via PropagationPolicy resources will be blocked from scheduling to the cluster if the policy doesn't specify matching taint tolerations.
> - Running applications that don't tolerate the taint will be automatically migrated by Karmada.
> 
> **Supported Taint Effects**
> 
> Currently, Karmada supports two taint effects:
> - **NoSchedule**: Affects the scheduling of new applications. When applied, newly created applications won't be scheduled to the tainted cluster.
> - **NoExecute**: Affects both scheduling and running applications. When applied, not only newly created applications won't be scheduled to the tainted cluster, but existing applications on the cluster will also be migrated.
> 
> **Automated Taint Management with ClusterTaintPolicy**
> 
> To streamline automated management of cluster taints, Karmada introduced the **ClusterTaintPolicy** resource in version v1.14. Administrators can define custom failure conditions for clusters and automate the "when and how" of cluster tainting based on specified conditions.

Reviewer: ```suggestion
Taints in a Cluster, specifically configured in the `.spec.taints` field of the `Cluster` object, have a significant 
impact on application scheduling and operation. For instance, if a cluster has a `NoSchedule` taint, it can prevent 
new applications from being scheduled onto that cluster. The effect of a taint on an application depends on the `Effect` 
configuration of the taint.

Karmada supports different types of taints based on their `Effect` values: 
- `NoSchedule`: Prevents new applications from scheduling to the cluster unless the application explicitly tolerates it.
- `NoExecute`: Prevents new scheduling **and evicts existing applications** from the cluster unless the application explicitly tolerates it.
Note that the eviction feature of NoExecute is disabled by default, to enable it, you need to use the 
`--enable-no-execute-taint-eviction=true` flag in the `karmada-controller-manager` and also explicitly enable the 
`Failover` feature gate.

This document explains how Karmada manages taints on Clusters, including
- Customized taint management based on cluster conditions.
- Taints managed by the Karmada system.
- Configuring application tolerations to bypass taint effects.
```

I prefer to go with some general introductions before diving into details, instead of categorizing by chapters right from the beginning.  Let me know does this makes sense to you.

> ## Taint Cluster By Conditions - ClusterTaintPolicy

Reviewer: ```suggestion
## Customized Taint Management
```
How about this title?

    - "True"
  - conditionType: NetworkAvailable
    operator: In
    statusValues:
    - "True"
  #...
```

The above example specifies that when the cluster's status conditions meet:
- `Ready` condition value is **True** **AND**
- `NetworkAvailable` condition value is **True**,

the policy will remove taints from the target cluster.

> Note: When you define both AddOnConditions and RemoveOnConditions in the ClusterTaintPolicy resource, AddOnConditions are evaluated before RemoveOnConditions.

### Configure Taints

`Taints` field specifies the taints that need to be added or removed on the cluster object which match with `TargetClusters`.

It can be configured as follows:

```yaml
apiVersion: policy.karmada.io/v1alpha1
kind: ClusterTaintPolicy
metadata:
  name: detect-cluster-notready
spec:
  #...
  taints:
  - key: not-ready
    effect: NoSchedule
  - key: not-ready
    effect: NoExecute
  #...
```

- When the cluster's status conditions simultaneously meet the `AddOnConditions`, the policy will add taints to the target cluster.
- When the cluster's status conditions simultaneously meet the `RemoveOnConditions`, the policy will remove taints from the target cluster.

> > Note: When you modified the Taints filed, the system will process the taints based on the latest value of Taints during the next condition-triggered execution, regardless of whether the taint has been added or removed.
> 
> > Note: When you delete the ClusterTaintPolicy, taints are NEVER automatically.

Reviewer: > taints are NEVER automatically.

NEVER automatically for what?

```suggestion
> Note: Modifying the taints field in a ClusterTaintPolicy (e.g., removing a taint definition) does not automatically 
clear existing taints applied by the policy. The controller only manages taints based on the updated configuration.

> Note: Deleting a ClusterTaintPolicy does not automatically clear existing taints applied by the policy. 
```

### Allow NoExecute Effect Taint

> Given that the **NoExecute** taint effect can evict all resources on a cluster — resulting in drastic impacts, the following parameter are adde to the `karmada-webhook` component to control whether users are allowed to configure taints with the **NoExecute** effect in `ClusterTaintPolicy`:

Reviewer: ```suggestion
Given the significant impact of NoExecute taints, which evict all applications that cannot tolerate the taint from the 
entire cluster, using ClusterTaintPolicy to configure NoExecute taints is disabled by default. This precaution prevents 
accidental widespread disruptions. To enable this capability, you must explicitly configure the karmada-webhook component
to allow NoExecute taint configurations:
```

- karmada-webhook
```
--allow-no-execute-taint-policy bool: Allows configuring taints with NoExecute effect in ClusterTaintPolicy. Given the impact of NoExecute, applying such a taint to a cluster may trigger the eviction of workloads that do not explicitly tolerate it, potentially causing unexpected service disruptions.
  This parameter is designed to remain disabled by default and requires careful evaluation by administrators before being enabled.
```

> In order to let `NoExecute` effect, after configure `--allow-no-execute-taint-policy=true` in the `karmada-webhook`, you also need to configure the following parameter in the `karmada-controller-manager` component:

Reviewer: ```suggestion
In order to let `NoExecute` effect, after configuring `--allow-no-execute-taint-policy=true` in the `karmada-webhook`, you also need to configure the following parameter in the `karmada-controller-manager` component:
```

This document explains how Karmada manages taints on Clusters, including
- Customized taint management based on cluster conditions.
- Taints managed by the Karmada system.
- Configuring application tolerations to bypass taint effects.

## Customized Taint Management

The `ClusterTaintPolicy` API allows administrators to define rules for applying or removing taints from clusters based on specific conditions. This is particularly useful for implementing advanced scheduling policies or failover strategies.

Cluster conditions can be determined not only by the Karmada system but also by any third-party system. This offers administrators great flexibility in defining cluster health criteria. For example, the Karmada system set the `Ready` condition based on its built-in health checks. However, a third-party system can introduce additional conditions, such as `SchedulerDown`, to indicate that the cluster scheduler is non-functional. When combined with `ClusterTaintPolicy`, these custom conditions can be used to prevent new applications from being scheduled onto affected clusters, providing more granular control over application placement and enhancing overall system reliability.

> ![](../../resources/userguide/failover/clustertaintpolicy.png)

Reviewer: I prefer to avoid using the `Ready` in this picture, as the taint based on that condition is usually not using ClusterTaintPolicy.

Author: Updated with SchedulerDown

The `ClusterTaintPolicy` is a Cluster-scoped resource, its apiVersion is `policy.karmada.io/v1alpha1`.

The following is an example of a `ClusterTaintPolicy` that adds or remove taints to the target cluster:

```yaml
apiVersion: policy.karmada.io/v1alpha1
kind: ClusterTaintPolicy
metadata:
  name: detect-cluster-notready
spec:
  targetClusters:
    clusterNames:
    - member1
    - member2
  addOnConditions:
  - conditionType: Ready
    operator: NotIn
    statusValues:
    - "True"
  removeOnConditions:
  - conditionType: Ready
    operator: In
    statusValues:
    - "True"
  taints:
  - key: not-ready
    effect: NoSchedule
```

> Next, let's take a look at the fields in the `ClusterTaintPolicy`.

Reviewer: ```suggestion
(Note: Add some explanation to the end of this sample.)
```
And can we use another condition for the example?

### Select a group of target clusters

`TargetClusters` field specifies the clusters that `ClusterTaintPolicy` needs to pay attention to. It uses the `ClusterAffinity` structure from the `PropagationPolicy` API to select target clusters. The usage is identical, so you can directly refer to [its documentation](https://karmada.io/docs/userguide/scheduling/resource-propagating#deploy-deployment-into-a-specified-set-of-target-clusters) for user-guide.

> In addition to this, there is one thing that needs to be noted, for clusters that no longer match the `TargetClusters`, the taints will be kept unchanged.

Reviewer: Need further explanation of why the cluster no longer matches.

### Add On Match Conditions

`AddOnConditions` field defines the conditions to match for triggering the `cluster-taint-policy-controller` to add taints on the cluster object. The match conditions are ANDed. If `AddOnConditions` is empty, no taints will be added.

It can be configured as follows:

```yaml

> apiVersion: "policy.karmada.io/v1alpha1"

Reviewer: ```suggestion
apiVersion: policy.karmada.io/v1alpha1
```

kind: "ClusterTaintPolicy"
metadata:
  name: "detect-cluster-notready"
spec:
  #...
  addOnConditions:
  - conditionType: "Ready"
    operator: "NotIn"
    statusValues:
    - "True"
  - conditionType: "NetworkAvailable"
    operator: "NotIn"
    statusValues:
    - "True"
  #...
```

The above example specifies that when the cluster's status conditions meet:
- `Ready` condition value is **not** True **AND**
- `NetworkAvailable` condition value is **not** True,

> the policy will add taints to the target cluster.

Author: We have already introduced the TargetClusters field above, and there is a full example at the beginning, so I think there is no need to explain taint and targetCluster here.

Karmada supports different types of taints based on their `Effect` values:
- `NoSchedule`: Prevents new applications from scheduling to the cluster unless the application explicitly tolerates it.
- `NoExecute`: Prevents new scheduling **and evicts existing applications** from the cluster unless the application explicitly tolerates it.

Note that the eviction feature of NoExecute is disabled by default, to enable it, you need to use the `--enable-no-execute-taint-eviction=true` flag in the `karmada-controller-manager` and also explicitly enable the `Failover` feature gate.

This document explains how Karmada manages taints on Clusters, including
- Customized taint management based on cluster conditions.
- Taints managed by the Karmada system.
- Configuring application tolerations to bypass taint effects.

## Customized Taint Management

> The `ClusterTaintPolicy` API allows administrators to define rules for applying or removing taints from clusters based on specific conditions. This is particularly useful for implementing advanced schedul ing policies or failover strategies.

Reviewer: ```suggestion
The `ClusterTaintPolicy` API allows administrators to define rules for applying or removing taints from clusters based on specific conditions. This is particularly useful for implementing advanced scheduling policies or failover strategies.
```

> Cluster conditions can be determined not only by the Karmada system but also by any third-party system. This offers administrators great flexibility in defining cluster health criteria. For example, the Karmada system set the `Ready` condition based on its built-in health checks. However, a third-party system can introduce additional conditions,
> 
> such as `SchedulerDown`, to indicate that the cluster scheduler is non-functional. When combined with `ClusterTaintPolicy`, these custom conditions can be used to prevent new applications from being scheduled onto affected clusters, providing more granular control over application placement and enhancing overall system reliability.

Reviewer: ```suggestion
Cluster conditions can be determined not only by the Karmada system but also by any third-party system. This offers administrators great flexibility in defining cluster health criteria. For example, the Karmada system sets the `Ready` condition based on its built-in health checks. However, a third-party system can introduce additional conditions, such as `SchedulerDown`, to indicate that the cluster scheduler is non-functional. When combined with `ClusterTaintPolicy`, these custom conditions can be used to prevent new applications from being scheduled onto affected clusters, providing more granular control over application placement and enhancing overall system reliability.
```
